### DISCUSSION TITLE
Remove Persisted State from localStorage?

### DISCUSSION THREAD

**Discussion by rohitpotato** [pmndrs/zustand#299] [Q&A]
I was wondering how to remove persisted state in local storage? I have tried the creating an action such as:

```
      removeAuthInfo: () =>
        set((state) => {
          return {
            ...omit(state, ["token", "lastTokenTimeStamp"]),
            isAuthenticated: false,
          };
        }),
    })
```

But this doesn't seem to remove anything from localStorage. Do we have to remove manually?

Entire store:

```
const useAuthStore = create(
  persist(
    (set) => ({
      isAuthenticated: false,
      setIsAuthenticated: (state) => set({ isAuthenticated: state }),
      lastTokenTimeStamp: null,
      setLastTokenTimeStamp: (lastTokenTimeStamp) =>
        set({ lastTokenTimeStamp }),
      token: null,
      setToken: (token) => set({ token }),
      removeAuthInfo: () =>
        set((state) => {
          return {
            ...omit(state, ["token", "lastTokenTimeStamp"]),
            isAuthenticated: false,
          };
        }),
    }),
    {
      name: localStorageKeys.TOKEN_INFORMATION,
      whitelist: ["token", "lastTokenTimeStamp"],
    }
  )
  ```

**Comment by AnatoleLucet** [comment]
Hi 👋, by default, Zustand merge the object given to `set()` with the current state. So I think for your code to work you'll need to call `set()` in "replace mode". You can do this by giving `true` as a second argument (e.g. `set((state) => ({ ... }), true)`).

But for this case, isn't setting `token` and `lastTokenTimeStamp` to `null` enough?

**Reply by rohitpotato** [reply]
> Hi 👋, by default, Zustand merge the object given to `set()` with the current state. So I think for your code to work you'll need to call `set()` in "replace mode". You can do this by giving `true` as a second argument (e.g. `set((state) => ({ ... }), true)`).
> 
> But for this case, isn't setting `token` and `lastTokenTimeStamp` to `null` enough?

The problem might occur for non-null values. For example state dependent on numerical values might cause visible fluctuations on the UI. Of course, one can use `useLayoutEffect` but causing breaks for the render doesn't seem like a very good idea. What do you think?

**Reply by AnatoleLucet** [reply]
I'm not sure what you mean. What problem?

> For example state dependent on numerical values might cause visible fluctuations on the UI. Of course, one can use useLayoutEffect but causing breaks for the render doesn't seem like a very good idea. What do you think?

Are you referring to #300 (`PersistGate`-like feature)?

**Reply by rohitpotato** [reply]
Ya kind of. Just to clarify, consider a state `volume` whose value is stored in the `localstorage` with value 0.2 but with a initial state of 0.5. Until `zustand` fetches the values from localstorage the volume will jump from 0.5 to 0.2 causing fluctuations in the UI. Correct me if i am missing something. 

This actually motivated me to open #300.

**Reply by AnatoleLucet** [reply]
> Correct me if i am missing something.

You're not. That's semi-intentional. Since we also wanted to support promise-based storages like React Native's `AsyncStorage`, we wrapped the store rehydration in an async function and `set()` the store once the storage's value has been retrieved ([see](https://github.com/pmndrs/zustand/blob/master/src/middleware.ts#L230)).

Currently, even non-promise-based storages are affected by this async rehydration. I assume this could be fixed by conditionally checking `storage.getItem()`'s returned value (e.g. pending promise or simple string). But I'm not sure if this change would cause some drawback (or not) to some use cases.

**Reply by osaxon** [reply]
> Hi 👋, by default, Zustand merge the object given to `set()` with the current state. So I think for your code to work you'll need to call `set()` in "replace mode". You can do this by giving `true` as a second argument (e.g. `set((state) => ({ ... }), true)`).
> 
> But for this case, isn't setting `token` and `lastTokenTimeStamp` to `null` enough?

Hi, this worked for me thank you. 

I was initially trying to do this:

`// Reset to default state
			reset: async () => {
				useCart.persist.clearStorage();
				set((state) => ({
					...initialState,
				}));
			},` 

the clearStorage method is documented in the zustand repo, but it didn't seem to work. I've removed this line and gone with your solution which works.

`
			reset: async () => {
				set(
					(state) => ({
						...initialState,
					}),
					// Zustand 'replace mode' overwrites existing state
					true
				);
			},
`

**Comment by bytesagar** [comment]
@AnatoleLucet  what if I dont want to persist my state. I dont want to show anything that i update in store in localstorage. How to achieve this?
